### PR TITLE
Circuit breaker

### PR DESCRIPTION
**Pull request by Pliner**
image

https://martinfowler.com/bliki/CircuitBreaker.html

```
        """
        failure_threshold: The failure threshold at which the circuit will break (a number between 0 and 1)
        break_duration: The duration the circuit will stay open before resetting
        minimum_throughput: How many actions must pass through the circuit breaker to come into action.
        sampling_sampling_duration: The duration when failure ratios are assessed.
        """
```

Example:

```
aio_request.DefaultCircuitBreaker[str, int](
        break_duration=1.0,
        sampling_duration=1.0,
        minimum_throughput=2,
        failure_threshold=0.5,
    )
```
In case of `requests count` >= `minimum throughput`(>=2) in `sampling period`(1 second) the circuit breaker will open if  `failed requests count`/`total requests count` >= `failure threshold`(50%).

### CODE REVIEW

File: aio_request/base.py [hunk 1]
```diff
@@ -24,6 +24,8 @@ class Header:
     X_REQUEST_PRIORITY = multidict.istr("X-Request-Priority")
     X_REQUEST_TIMEOUT = multidict.istr("X-Request-Timeout")
     X_SERVICE_NAME = multidict.istr("X-Service-Name")
+    X_DO_NOT_RETRY = multidict.istr("X-Do-Not-Retry")
+    X_CIRCUIT_BREAKER = multidict.istr("X-Circuit-Breaker")
```
Comment (Pliner) [inline aio_request/base.py:28]: To be able to check in the client that request was produced by Circuit Breaker

File: aio_request/base.py [hunk 2]
```diff
@@ -24,6 +24,8 @@ class Header:
     X_REQUEST_PRIORITY = multidict.istr("X-Request-Priority")
     X_REQUEST_TIMEOUT = multidict.istr("X-Request-Timeout")
     X_SERVICE_NAME = multidict.istr("X-Service-Name")
+    X_DO_NOT_RETRY = multidict.istr("X-Do-Not-Retry")
```
Comment (Pliner) [inline aio_request/base.py:27]: Do tell a strategy to stop sending requests and accept a response as is

File: aio_request/circuit_breaker.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
+import abc
+import collections
+import enum
+import time
+from typing import Awaitable, Callable, DefaultDict, Generic, Mapping, Optional, TypeVar
+
+
+class CircuitState(str, enum.Enum):
+    OPENED = "OPENED"
```
Comment (spaceorc) [inline aio_request/circuit_breaker.py:9]: Lets maybe name it OPEN and HALF_OPEN as in article?

File: aio_request/pipeline.py [hunk 4]
```diff
@@ -31,7 +34,16 @@ async def execute(
         ...
 
 
-class LowTimeoutRequestModule(RequestModule):
+class ByPassModule(RequestModule):
```
Comment (spaceorc) [inline aio_request/pipeline.py:37]: Also suggestion: let's name it BypassModule (small letter 'p')

File: tests/test_circuit_breaker.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    assert await circuit_breaker.execute(scope="scope", operation=do(500), fallback=503, is_successful=is_200) == 500
+    assert await circuit_breaker.execute(scope="scope", operation=do(500), fallback=503, is_successful=is_200) == 503
+
+    await asyncio.sleep(1)
+
+    assert await circuit_breaker.execute(scope="scope", operation=do(200), fallback=503, is_successful=is_200) == 200
+    assert circuit_breaker.state == {"scope": aio_request.CircuitState.CLOSED}
+
+
+async def test_circuit_breaker_should_close_after_failure() -> None:
```
Comment (spaceorc) [inline tests/test_circuit_breaker.py:71]: should_open?

File: aio_request/circuit_breaker.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    def __init__(
+        self,
+        *,
+        break_duration: float,
+        failure_threshold: float,
+        minimum_throughput: float,
+        sampling_duration: float,
+    ):
+        """
+        failure_threshold: The failure threshold at which the circuit will break (a number between 0 and 1)
```
Comment (spaceorc) [inline aio_request/circuit_breaker.py:131]: Should we check all this limits? E.g. if user passes sampling_duration=0 our RollingCircuitBreakerMetrics will fail...

File: aio_request/circuit_breaker.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        break_duration: The duration the circuit will stay open before resetting
+        minimum_throughput: How many actions must pass through the circuit breaker to come into action.
+        sampling_sampling_duration: The duration when failure ratios are assessed.
```
Comment (spaceorc) [inline aio_request/circuit_breaker.py:134]: typo: sampling_sampling

File: aio_request/circuit_breaker.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        self,
+        scope: TScope,
+        operation: Callable[[], Awaitable[TResult]],
+        fallback: TResult,
+        is_successful: Callable[[TResult], bool],
+    ) -> TResult:
+        if not self._is_executable(scope):
+            return fallback
+
+        result = await operation()
```
Comment (spaceorc) [inline aio_request/circuit_breaker.py:155]: I'm not sure that we should deal with it somehow, but while we are inside this await there can occur one or more calls to execute with the same scope (e.g. operation can work a lot of time).

**Review by spaceorc — CHANGES_REQUESTED** [review]

**Review by spaceorc — APPROVED** [review]

**Pull request merged**